tilbakekreving/application/src/main/kotlin/tilbakekreving/application/service/kravgrunnlag/AnnullerKravgrunnlagService.kt

```diff
@@ -0,0 +1,126 @@
+import tilbakekreving.domain.KanAnnullere
+import tilbakekreving.domain.TilbakekrevingsbehandlingRepo
+import tilbakekreving.domain.kravgrunnlag.AnnullerKravgrunnlagCommand
+import tilbakekreving.domain.kravgrunnlag.Kravgrunnlag
+import tilbakekreving.domain.kravgrunnlag.Kravgrunnlagstatus
+import tilbakekreving.domain.kravgrunnlag.påsak.KravgrunnlagStatusendringPåSakHendelse
+import tilbakekreving.domain.kravgrunnlag.repo.AnnullerKravgrunnlagStatusEndringMeta
+import tilbakekreving.domain.kravgrunnlag.repo.KravgrunnlagRepo
+import tilbakekreving.domain.vedtak.KunneIkkeAnnullerePåbegynteVedtak
+import tilbakekreving.domain.vedtak.Tilbakekrevingsklient
+import tilgangstyring.application.TilgangstyringService
+import tilgangstyring.domain.IkkeTilgangTilSak
+import java.time.Clock
+
+class AnnullerKravgrunnlagService(
+    private val tilgangstyring: TilgangstyringService,
+    private val tilbakekrevingsbehandlingRepo: TilbakekrevingsbehandlingRepo,
+    private val sakService: SakService,
+    private val kravgrunnlagRepo: KravgrunnlagRepo,
+    private val tilbakekrevingsklient: Tilbakekrevingsklient,
+    private val sessionFactory: SessionFactory,
+    private val clock: Clock,
+) {
+    private val log = LoggerFactory.getLogger(this::class.java)
+
+    fun annuller(command: AnnullerKravgrunnlagCommand): Either<KunneIkkeAnnullereKravgrunnlag, Pair<Kravgrunnlag, AvbruttTilbakekrevingsbehandling?>> {
+        tilgangstyring.assertHarTilgangTilSak(command.sakId).onLeft {
+            return KunneIkkeAnnullereKravgrunnlag.IkkeTilgang(it).left()
+        }
+        val sak = sakService.hentSak(command.sakId).getOrElse {
+            throw IllegalStateException("Kunne ikke oppdatere kravgrunnlag for tilbakekrevingsbehandling, fant ikke sak. Command: $command")
+        }
+        if (sak.versjon != command.klientensSisteSaksversjon) {
+            log.info("Oppdater kravgrunnlag - Sakens versjon (${sak.versjon}) er ulik saksbehandlers versjon. Command: $command")
+        }
+
+        val tilbakekrevingsbehandlingHendelser = tilbakekrevingsbehandlingRepo.hentForSak(command.sakId)
+        val uteståendeKravgrunnlagPåSak = tilbakekrevingsbehandlingHendelser.hentUteståendeKravgrunnlag()
+            ?: return KunneIkkeAnnullereKravgrunnlag.SakenHarIkkeKravgrunnlagSomKanAnnulleres.left()
+
+        val kravgrunnlagOgBehandling =
+            tilbakekrevingsbehandlingHendelser.hentKravgrunnlagOgBehandlingFor(command.kravgrunnlagHendelseId)
+                ?: return KunneIkkeAnnullereKravgrunnlag.FantIkkeKravgrunnlag.left()
+
+        if (uteståendeKravgrunnlagPåSak.hendelseId != kravgrunnlagOgBehandling.first.hendelseId) {
+            return KunneIkkeAnnullereKravgrunnlag.InnsendtHendelseIdErIkkeDenSistePåSaken.left()
+        }
+
+        val avbruttHendelseOgBehandling = kravgrunnlagOgBehandling.second.let {
+            if (it != null) {
+                (kravgrunnlagOgBehandling.second as? KanAnnullere)?.annuller(
```

Comment:
kan vel bruke it istedenfor kravgrunnlagOgBehandling.second

Reply:
Henger igjen fra da jeg glemte å pushe etter å ha sett over